# Homework #3
**Name:** Hanwen Li
**Email:** lihanwen@seas.upenn.edu  

In [2]:
import numpy as np
def get_divider(id):
    print('**************************************************** Problem {} ****************************************************'.format(id))

## Problem #1
See the pdf

## Problem #2

In [10]:
class Problem2():
    def __init__(self):
        get_divider(2)
        self.X = None
        self.Y = None
        self.P = None
        self.L = None


    def get_cordinate(self):
        self.X = np.random.uniform()
        self.Y = np.random.uniform()

    def get_distance(self):
        return (self.X**2+self.Y**2)**0.5
    
    def get_probability(self, L):
        if L <= 1:
            self.P =  L*L*np.pi/4
        else:
            self.P = np.pi/4*L*L + (L*L-1)**0.5 - L*L*np.arctan((L*L-1)**0.5)

    def simulation(self, num):
        self.L = np.random.uniform(low=0, high=2)**0.5
        self.get_probability(self.L)
        distance = []
        lay_in = []
        for i in range(num):
            self.get_cordinate()
            distance.append(self.get_distance())
            if self.get_distance() <= self.L:
                lay_in.append(1)
            else:
                lay_in.append(0)

        return lay_in, distance


In [11]:
problem2 = Problem2()
for i in range(10):
    lay_in, distance = problem2.simulation(3000)
    print("The result of simulation is {:.3f}, while the result of calculation is {:.3f}, and the average distance is {:.3f}".format(np.mean(lay_in), problem2.P,np.mean(distance)))

**************************************************** Problem 2 ****************************************************
The result of simulation is 0.980, while the result of calculation is 0.983, and the average distance is 0.764
The result of simulation is 0.802, while the result of calculation is 0.803, and the average distance is 0.764
The result of simulation is 0.614, while the result of calculation is 0.617, and the average distance is 0.772
The result of simulation is 0.534, while the result of calculation is 0.537, and the average distance is 0.768
The result of simulation is 0.392, while the result of calculation is 0.397, and the average distance is 0.763
The result of simulation is 0.855, while the result of calculation is 0.856, and the average distance is 0.772
The result of simulation is 0.866, while the result of calculation is 0.871, and the average distance is 0.764
The result of simulation is 0.948, while the result of calculation is 0.947, and the average distance is 0.

### Result
The result of simulation is 0.980, while the result of calculation is 0.983, and the average distance is 0.764

The result of simulation is 0.802, while the result of calculation is 0.803, and the average distance is 0.764

The result of simulation is 0.614, while the result of calculation is 0.617, and the average distance is 0.772

The result of simulation is 0.534, while the result of calculation is 0.537, and the average distance is 0.768

The result of simulation is 0.392, while the result of calculation is 0.397, and the average distance is 0.763

The result of simulation is 0.855, while the result of calculation is 0.856, and the average distance is 0.772

The result of simulation is 0.866, while the result of calculation is 0.871, and the average distance is 0.764

The result of simulation is 0.948, while the result of calculation is 0.947, and the average distance is 0.769

The result of simulation is 0.964, while the result of calculation is 0.963, and the average distance is 0.759

The result of simulation is 0.993, while the result of calculation is 0.995, and the average distance is 0.767

## Problem #3
See the pdf

## Problem #4
See the pdf

## Problem #5

In [12]:
class Problem5():
    def __init__(self):
        get_divider(5)
        self.A = None
        self.B = None
        self.C = None

    def get_random_variables(self, l):
        self.A =  np.random.exponential(l)
        self.B = np.random.exponential(l)
        self.C = np.random.exponential(l)

    def get_discriminant(self):
        return self.B**2 - 4*self.A*self.C
    
    def simulate(self, num, l):
        real = []
        for i in range(num):
            self.get_random_variables(l)
            if self.get_discriminant() >= 0:
                real.append(1)
            else:
                real.append(0)
        return real

In [13]:
problem = Problem5()

for l in [0.5, 1, 2, 5, 10]:
    real = problem.simulate(5000, l)
    print("The result of simulation is {:.3f}".format(np.mean(real)))

**************************************************** Problem 5 ****************************************************
The result of simulation is 0.340
The result of simulation is 0.327
The result of simulation is 0.337
The result of simulation is 0.335
The result of simulation is 0.327


### Result
The result of simulation is 0.340

The result of simulation is 0.327

The result of simulation is 0.337

The result of simulation is 0.335

The result of simulation is 0.327

## problem #6

In [14]:
class Problem6():
    def __init__(self):
        get_divider(6)
        self.P = {'A':None, 'B':None}

    def set_probability(self):
        self.P['A'] = np.random.uniform()
        self.P['B'] = np.random.uniform()

    def equation(self, question='a'):
        if question == 'a':
            return self.P['A']/(1-(1-self.P['A'])*(1-self.P['B']))
        else:
            return (2-self.P['A'])/(self.P['A']+self.P['B']-self.P['A']*self.P['B'])
        
    def roll(self, role='A'):
        if role == 'A':
            return np.random.choice([1, 0], p=[self.P['A'], 1 - self.P['A']])
        else:
            return np.random.choice([1, 0], p=[self.P['B'], 1 - self.P['B']])
        
    def simulate(self, num):
        self.set_probability()
        rolls = []
        a_wins = []
        for i in range(num):
            rolls.append(0)
            while True:
                rolls[i] += 1
                if self.roll('A') == 1:
                    a_wins.append(1)
                    break
                else:
                    rolls[i] += 1
                    if self.roll('B') == 1:
                        a_wins.append(0)
                        break

        return rolls, a_wins
            

In [18]:
problem = Problem6()

for i in range(5):
    rolls, a_wins = problem.simulate(5000)
    print("The result of simulation is rolls : {:.3f} and A wins rolls, a_wins: {:.3f}".format(np.mean(rolls), np.mean(a_wins)))
    print("The result of calculation is rolls : {:.3f} and A wins rolls, a_wins: {:.3f}".format(problem.equation('b'), problem.equation('a')))
    print('-'*50)

**************************************************** Problem 6 ****************************************************
The result of simulation is rolls : 1.199 and A wins rolls, a_wins: 0.909
The result of calculation is rolls : 1.207 and A wins rolls, a_wins: 0.911
--------------------------------------------------
The result of simulation is rolls : 1.237 and A wins rolls, a_wins: 0.829
The result of calculation is rolls : 1.234 and A wins rolls, a_wins: 0.831
--------------------------------------------------
The result of simulation is rolls : 1.904 and A wins rolls, a_wins: 0.497
The result of calculation is rolls : 1.897 and A wins rolls, a_wins: 0.505
--------------------------------------------------
The result of simulation is rolls : 2.978 and A wins rolls, a_wins: 0.030
The result of calculation is rolls : 2.929 and A wins rolls, a_wins: 0.030
--------------------------------------------------
The result of simulation is rolls : 2.766 and A wins rolls, a_wins: 0.664
The result

### Result
The result of simulation is rolls : 1.206 and A wins rolls, a_wins: 0.979

The result of calculation is rolls : 1.215 and A wins rolls, a_wins: 0.976
**************************************************
The result of simulation is rolls : 3.061 and A wins rolls, a_wins: 0.083

The result of calculation is rolls : 3.052 and A wins rolls, a_wins: 0.083
**************************************************
The result of simulation is rolls : 1.987 and A wins rolls, a_wins: 0.113

The result of calculation is rolls : 1.991 and A wins rolls, a_wins: 0.117
**************************************************
The result of simulation is rolls : 1.966 and A wins rolls, a_wins: 0.582

The result of calculation is rolls : 1.947 and A wins rolls, a_wins: 0.582
**************************************************
The result of simulation is rolls : 4.196 and A wins rolls, a_wins: 0.774

The result of calculation is rolls : 4.208 and A wins rolls, a_wins: 0.773
**************************************************

## Simulation Problem #6

In [3]:
class SimulationProblem:
  def __init__(self):
    get_divider('Simulation')
    self.max_capacity = 250
    self.stock = 250
    self.demand = None
    self.supply = 0
    self.profit = 10
    self.storage_cost = 6

  def set_demand(self):
    self.demand = round(np.random.uniform(low=24.5, high=50.5))

  def set_supply(self):
    self.supply = round(np.random.uniform(low=0.5, high=5.5))

  def get_daily_profit(self):
    profit = 0
    if self.demand > self.stock:
      profit += self.stock*self.profit
      profit -= (self.demand-self.stock)*self.profit
      self.stock = 0
    else:
      profit += self.demand*self.profit
      self.stock -= self.demand

    return profit

  def get_storage_cost(self, arrive=False):
    if arrive:
      self.stock += self.max_capacity
      self.supply = -1

    return -(self.stock - self.max_capacity)*self.storage_cost if self.stock > self.max_capacity else 0

  def simulate(self, num, l):
    profits = []
    for i in range(num):
      self.set_demand()
      profit = self.get_daily_profit()
      if self.supply > 1:
        self.supply -= 1
        profit += self.get_storage_cost()
      elif self.supply == 1:
        profit += self.get_storage_cost(arrive=True)
      else:
        if self.stock < l*self.max_capacity:
          self.set_supply()
          profit += self.get_storage_cost()
      profits.append(profit)

    return profits

In [6]:
problem = SimulationProblem()

for i in range(21):
  profits = problem.simulate(10000, i/20)
  print("The result of simulation with lambda {} is {:.1f}".format(i/20, np.mean(profits)))

**************************************************** Problem Simulation ****************************************************
The result of simulation with lambda 0.0 is -375.0
The result of simulation with lambda 0.05 is 130.6
The result of simulation with lambda 0.1 is 149.2
The result of simulation with lambda 0.15 is 167.6
The result of simulation with lambda 0.2 is 189.1
The result of simulation with lambda 0.25 is 210.6
The result of simulation with lambda 0.3 is 222.2
The result of simulation with lambda 0.35 is 239.8
The result of simulation with lambda 0.4 is 258.2
The result of simulation with lambda 0.45 is 273.9
The result of simulation with lambda 0.5 is 283.1
The result of simulation with lambda 0.55 is 293.6
The result of simulation with lambda 0.6 is 300.7
The result of simulation with lambda 0.65 is 306.3
The result of simulation with lambda 0.7 is 307.6
The result of simulation with lambda 0.75 is 308.4
The result of simulation with lambda 0.8 is 305.0
The result of si

### Result
The max profit is achieved when lambda is in [0.65, 0.7]